test_runner/fixtures/benchmark_fixture.py

```diff
@@ -411,14 +414,30 @@ def record_pageserver_writes(
 
 
 @pytest.fixture(scope="function")
-def zenbenchmark(record_property: Callable[[str, object], None]) -> Iterator[NeonBenchmarker]:
+def zenbenchmark(
+    request: FixtureRequest,
+    record_property: Callable[[str, object], None],
+) -> Iterator[NeonBenchmarker]:
     """
     This is a python decorator for benchmark fixtures. It contains functions for
     recording measurements, and prints them out at the end.
     """
     benchmarker = NeonBenchmarker(record_property)
     yield benchmarker
 
+    with tempfile.NamedTemporaryFile(mode="w") as tmp:
+        results = {}
+        for _, recorded_property in request.node.user_properties:
+            name = recorded_property["name"]
+            value = recorded_property["value"]
+            unit = recorded_property["unit"]
+            results[name] = f"{value} {unit}"
+
+        json.dump(results, tmp, indent=2)
+        tmp.flush()
+
+        allure.attach.file(tmp.name, "benchmarks-results.json", allure.attachment_type.JSON)
+
```

Comment:
Why side-step through the NamedTemporaryFile instead of straight-forward `allure.attach()`?

https://allurereport.org/docs/pytest-reference/


Also, I don't understand how pytest's `user_properties` learn about the `zenbenchmark.record()'s name` `value` and `unit`, but, I'm sure you know what you're doing.

Reply:
> Why side-step through the NamedTemporaryFile instead of straight-forward allure.attach()?

Ah, good catch! Will switch it to `allure.attach` instead.

> Also, I don't understand how pytest's user_properties learn about the zenbenchmark.record()'s name value and unit, but, I'm sure you know what you're doing.

We do it in a bit unusual way, we use [`record_property`](https://docs.pytest.org/en/7.1.x/how-to/output.html#record-property) (which is a part of a JUnit formatter); it allows to get recoded properties though [`user_properties`](https://docs.pytest.org/en/7.1.x/reference/reference.html#pytest.Item.user_properties).
Probably it's leftovers from those days we used junit formatter, but it still works for benchmarking results 🤷